src/transport/connection.rs

```diff
@@ -14,11 +14,14 @@ use tokio::{
 
 use Conditional;
 use transport::{AddrInfo, BoxedIo, GetOriginalDst, tls};
+use indexmap::IndexSet;
 
-pub struct BoundPort {
-    inner: std::net::TcpListener,
+pub struct BoundPort<G = ()> {
+    inner: Option<std::net::TcpListener>,
```

Comment:
Out of curiousity, why does this need to be wrapped inside an option?

Reply:
Yeah, so that's because we need to [move](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html#ways-variables-and-data-interact-move) the socket into [`tokio::net::TcpListener::from_std`](https://tokio-rs.github.io/tokio/tokio/net/tcp/struct.TcpListener.html#method.from_std) (which takes a `std::net::TcpListener` _by value_) here:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L230
but, we now call the new `new_conn` method on a reference to the struct (`&self`) in the closure passed to `Stream::and_then` here:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L239-L248

Since the listener can no longer be moved out of the struct by deconstructing it, as the struct still needs to be alive when the closure is invoked, we now use [`Option::take`](https://doc.rust-lang.org/std/option/enum.Option.html#method.take), which moves the value out of the `Option`, replacing it with `None`. We can use this to move the `TcpListener` out of the `BoundPort` struct without dropping the `BoundPort`:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L222-L223
since `Option::take` takes the receiver by a _mutable reference_ (`&mut self`). 

This is a fairly common pattern in Rust when dealing with similar situations.